middleware/metrics.go

```diff
@@ -0,0 +1,218 @@
+package middleware
+
+// Ported from Goji's middleware, source:
+// https://github.com/zenazn/goji/tree/master/web/middleware
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+)
+
+// Metric is a composition of various statistics from a request, such as its
+// execution time and the response status.
+type Metric struct {
+	RequestID     string
+	Method        string
+	URL           string
+	RemoteAddr    string
+	Status        int
+	BytesWritten  int
+	ExecutionTime time.Duration
+}
+
+// MetricsHandler is a Metric receiver.
+type MetricsHandler interface {
+	Add(metric Metric)
+}
+
+// Metrics return a middleware which will use the given MetricsHandler in
+// order to publish new statistics.
+func Metrics(handler MetricsHandler) func(next http.Handler) http.Handler {
```

Comment:
I feel like Metrics is more specific word for the application data/metrics (where data would contain session, user_id, ip, referer, user_agent, context etc.), not just the Request/Response data.